unittest/Dialect/TritonGPU/DialectTest.cpp

```diff
@@ -0,0 +1,509 @@
+
+  std::minstd_rand rng(/*seed=*/0);
+  auto randPow2Vec = [&](int rank, int maxPow2) {
+    SmallVector<unsigned> ret;
+    for (int i = 0; i < rank; i++) {
+      int pow2 = std::uniform_int_distribution<unsigned>(0, maxPow2)(rng);
+      if (pow2 == maxPow2 && maxPow2 > 0) {
+        maxPow2--;
+      }
+      ret.push_back(1 << pow2);
+    }
+    return ret;
+  };
+
+  int numSuccess = 0;
+  for (int i = 0; i < numTests; i++) {
+    SCOPED_TRACE("iteration " + std::to_string(i));
+    int rank = std::uniform_int_distribution<int>(1, 4)(rng);
+
+    SmallVector<int64_t> srcShape(
+        convertType<int64_t>(randPow2Vec(rank, /*maxPow2=*/4)));
+    SmallVector<int64_t> dstShape = srcShape;
+    std::shuffle(dstShape.begin(), dstShape.end(), rng);
+
+    // Optionally merge some dimensions in dst.
+    for (int i = 1; i < dstShape.size(); i++) {
+      if (std::uniform_real_distribution<float>(0, 1)(rng) > 1.0 / rank) {
+        dstShape[i - 1] *= dstShape[i];
+        dstShape.erase(dstShape.begin() + i);
+        i--;
+      }
+    }
+
+    SmallVector<unsigned> sizePerThread = randPow2Vec(rank, /*maxPow2=*/3);
+    SmallVector<unsigned> threadsPerWarp = randPow2Vec(rank, /*maxPow2=*/3);
+    SmallVector<unsigned> warpsPerCTA = randPow2Vec(rank, /*maxPow2=*/3);
+
+    SmallVector<unsigned> order(llvm::to_vector(llvm::seq<unsigned>(rank)));
+    std::shuffle(order.begin(), order.end(), rng);
+
+    auto ctaLayout = CTALayoutAttr::get(
+        &ctx, SmallVector<unsigned>(rank, 1), SmallVector<unsigned>(rank, 1),
+        llvm::to_vector(llvm::reverse(llvm::seq<unsigned>(rank))));
+
+    auto srcTy = RankedTensorType::get(
+        srcShape, FloatType::getF32(&ctx),
+        BlockedEncodingAttr::get(&ctx, sizePerThread, threadsPerWarp,
+                                 warpsPerCTA, order, ctaLayout));
+    auto dstTy = RankedTensorType::get(dstShape, FloatType::getF32(&ctx));
+
+    bool couldReshape = false;
+    testReshape(srcTy, dstTy, /*expectedDstEnc=*/std::nullopt,
+                /*expectSuccess=*/std::nullopt, inferLayout,
+                /*longErrors=*/false, &couldReshape);
+    if (couldReshape)
+      numSuccess++;
+  }
+
+  // We don't expect or want 100% success, but if only a tiny fraction of tests
+  // actually exercise the successful reshape logic, then that gives us bad
+  // coverage.  I'm currently getting 35% success, which seems good enough,
+  // especially since the successful cases take a lot longer to run because of
+  // the MultiIdx checks (so we're spending most of our time on successful
+  // cases, even if they're only 1/3 of the iterations).
+  //
+  // Run ctest with --verbose to see this output.  For example:
+  //   $ cd python/build/cmake.blah.blah
+  //   $ ninja
+  //   $ $(git rev-parse --show-toplevel)/.venv/bin/ctest --verbose
+  printf("Fuzz success rate: %d/%d = %.2f%%\n", numSuccess, numTests,
+         100.0 * numSuccess / numTests);
```

Comment:
Does this increase CI time significantly?

Reply:
> We don't expect or want 100% success

Are those failed cases all belong to a specific category like `Expected reshape to succeed, but it didn't!`, or you are counting to all possible failures?

Reply:
On my laptop the fuzz test takes 0.25s, taking the C++ tests from 1.5s to 1.75s.

Reply:
> Are those failed cases all belong to a specific category like Expected reshape to succeed, but it didn't!, or you are counting to all possible failures?

We don't expect any of these reshapes to succeed (because the fuzzer doesn't know which are valid src encodings).

I'm counting all of the possible ways that we might reject a reshape of `(srcShape, srcEnc, dstShape)` as impossible.

Reply:
> On my laptop the fuzz test takes 0.25s, taking the C++ tests from 1.5s to 1.75s.

Not bad :)